[FRONTEND] Create creditCardList component for checkout page

ozdenizdolu opened the pull request:
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Description

I have added the credit card list component to the checkout page, created `checkoutPage.js`, added it to the route `./checkout`

# Screenshots


![1](https://user-images.githubusercontent.com/32651175/101283308-8cdb7680-37ea-11eb-97cf-d93f296992ae.jpg)
![2](https://user-images.githubusercontent.com/32651175/101283309-8d740d00-37ea-11eb-8ac2-81652de27990.jpg)
![3](https://user-images.githubusercontent.com/32651175/101283311-8d740d00-37ea-11eb-916c-3137b35d23b6.jpg)
![4](https://user-images.githubusercontent.com/32651175/101283307-8c42e000-37ea-11eb-9478-e2544c2fb506.jpg)


# Test results

Tested by hand. Seems like working.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks

- It still hasn't been integrated with the backend. Some data formatting will be needed. (EDIT: Partially solved)
- Edit & Add card forms only validate for 'required inputs' not for the format of the inputs. (EDIT: Solved)
- There might be visual bugs if there are too many cards. (EDIT: not as buggy as expected)

Closes #222

Commit:
Created checkout page & added credit card list
- Still not complete.

Commit:
Completed the CreditCardList(except some details)
- Created edit, add a card modals
- Created delete popup

MehdiSaffar commented on ecommerce/frontend/src/components/pages/CheckoutPage.js:
```diff
@@ -0,0 +1,5 @@
+import { CreditCardList } from "../checkout/CreditCardList"
+
+export const CheckoutPage = () => {
+    return <CreditCardList></CreditCardList>
```
`<CreditCardList/>` instead.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -0,0 +1,50 @@
+import './CreditCardList.less'
+
+export const CreditCardList = () => {
+    const [cardList, setCardList] = useState([
```
Maybe include some draft code for requesting cards from backend? And having some loading indicator too?

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -0,0 +1,50 @@
+export const CreditCardList = () => {
+    const [cardList, setCardList] = useState([
+        {
+            cardProps: {
```
Instead of using the format of the credit card library use the actual format from backend and have a function that formats it from backend format to library format

MehdiSaffar committed:
formatting

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -0,0 +1,57 @@
+                    dataSource={cardList}
+                    renderItem={item => (
+                        <List.Item>
+                            <CreditCard {...item} selected={selectedCard === item.cardId} onSelect={onCardSelect} />
```
expect `card` prop instead

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -0,0 +1,57 @@
+    const [selectedCard, setSelectedCard] = useState()
+    const [addVisible, setAddVisible] = useState(false)
+
+    const onCardSelect = value => setSelectedCard(value)
```
replace `value` with `cardId` and maybe rename `selectedCard` to `selectedCardId`

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCard.js:
```diff
@@ -0,0 +1,61 @@
+                    &nbsp;
+                    <Popconfirm
+                        title="Delete this card?"
+                        onConfirm={() => {}}
```
Maybe write some draft functions with the usual try catch thing and showing success or error notifications, maybe to "mock" the request put a sleep delay

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+    setVisible = () => {},
+}) => {
+    /* This might not remounted after closing and opening again, be careful */
+    const [cvc, setCvc] = useState(cardProps.cvc)
+    const [expiry, setExpiry] = useState(cardProps.expiry)
+    const [name, setName] = useState(cardProps.name)
+    const [number, setNumber] = useState(cardProps.number)
+    const [cardNickname, setCardNickname] = useState(cardName)
```
use `form.getFieldsValue()` in a `onFieldsChange` callback from form like I showed you

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+import 'react-credit-cards/es/styles-compiled.css'
+import { Modal, Form, Input } from 'antd'
+
+export const EditCardModal = ({
+    cardProps = {
+        cvc: '',
+        expiry: '',
+        number: '',
+        name: '',
+    },
+    cardId = {},
+    cardName = '',
+    visible = false,
+    setVisible = () => {},
+}) => {
```
Instead of having EditCardModal and AddCardModal have a single parent component that takes the id of the card and the "editing" mode as "add" or "edit" and then pass down the initial value (if there is) and an onSubmit callback to process the new values. Also instead of taking setVisibile take onOk and onCancel instead.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+                form={form}
+                name="editCardForm"
+                onFinish={() => {}}
+                onFinishFailed={() => {}}
```
no need for onFinishFailed

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+                layout="horizontal"
+                form={form}
+                name="editCardForm"
+                onFinish={() => {}}
```
this will be passed up to the parent component as onSubmit (or onFinish however you want to call it)

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+                {...formItemLayout}
+                layout="horizontal"
+                form={form}
+                name="editCardForm"
```
change the name

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+                    ]}>
+                    <Input
+                        onFocus={handleInputFocus}
+                        onChange={val => {
+                            setName(val.target.value)
+                        }}
```
Remove all onChanges

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -0,0 +1,157 @@
+                    name="expiry"
+                    label="Expiry Date"
+                    rules={[
+                        {
+                            required: true,
+                            message: 'Please input your card expiry date!',
+                        },
```
Add stricter validation on all fields. Also assist the user by adding `/` in SKT field for example and `-` in card number

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/AddCardModal.js:
```diff
@@ -0,0 +1,157 @@
+                    />
+                </Form.Item>
+                <Form.Item
+                    name="name"
+                    label="Name"
+                    rules={[
+                        {
+                            required: true,
+                            message: 'Please input your card name!',
+                        },
+                    ]}>
+                    <Input
+                        onFocus={handleInputFocus}
+                        onChange={val => {
+                            setName(val.target.value)
+                        }}
+                        name="name"
+                    />
```
Make the distinction clearer by calling Card Name and Owner Name

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCard.less:
```diff
@@ -0,0 +1,22 @@
+.creditcard-container {
+    padding: 16px;
+    display: inline-block;
+    &__selected {
+        border: 2px solid gainsboro;
+    }
```
border moves card by 4px in total, so assign transparent border even when not selected

MehdiSaffar reviewed (changes requested):
We went through this PR together in a google meet and these are the main points:

1. Have a single modal container with two modes add and edit
2. Have the card appear at the top
3. Have CCV and SKT appear next to each other
4. Strict data validation
5. Show loading spinner on Edit/Add button
6. Show notifications when success or error. If error don't close the modal
7. Refresh the entire list on success (to get the new cards)
8. Border 2px thing
9. itemKey in list
10. Show loading spinner when loading the cards in the list
11. If multiple cards presented, decide if getting default card from backend or simply choosing the first card in the list
12. When list empty show a friendlier message (like Add a payment method to checkout)

Other than that I think you did an excellent job and the card animation looks really good :+1:

Commit:
Done many changes according to review
- EditCardModal and AddCardModal is more modular now
- Added notifications and spinners for every action
- Added a data formatter to utils.js

Commit:
Added stricter field validations to card edit
- Also did some minor fixes here and there.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/AddCardModal.js:
```diff
@@ -1,157 +1,47 @@
 import { useState } from 'react'
-import Cards from 'react-credit-cards'
 import 'react-credit-cards/es/styles-compiled.css'
```
is this import required by the library? I mean do they ask to import it this way?

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/AddCardModal.js:
```diff
@@ -1,157 +1,47 @@
+            console.log(fields)
+            onSuccessfulAdd();
+            onCancel();
+            notification.success({ message: 'You have successfully added your credit card!' })
```
The ordering is wrong here. 

First get the form values, then set loading true, then simulate sending request with sleep, then show notification, then in the `finally { ... }` close set loading to false.

Reason we do it in finally is not to forget to set the loading back to false if an error happens. Also could you add `console.error(error)` in the catch?

Other remark is not to call multiple callbacks (onSuccessfulAdd and onCancel). There should be only one.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CardModalInner.js:
```diff
@@ -0,0 +1,179 @@
+    mode = 'add'
+}) => {
+
+    const [cardState, setCardState] = useState(
```
Okay I was looking for where the `mode` prop will popup in this PR :D

I would say move the `mode` prop to the parent, which simply passes down initialValues. so if `edit` mode it will pass the initial values, if `add` mode it will pass an empty initialValues. The CardModalInner should be rather a dumb component.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CardModalInner.js:
```diff
@@ -0,0 +1,179 @@
+    )
+    const [focused, setFocused] = useState('')
+
+    const formItemLayout = {
+        labelCol: {
+            xs: {
+                span: 24,
+            },
+            sm: {
+                span: 6,
+            },
+        },
+        wrapperCol: {
+            xs: {
+                span: 24,
+            },
+            sm: {
+                span: 18,
+            },
+        },
+    }
```
this object will be recreated on every render. since it is constant I'd say move it outside the component

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CardModalInner.js:
```diff
@@ -0,0 +1,179 @@
+    const handleInputFocus = val => setFocused(val.target.name)
+
+    const onFieldsChange = () => setCardState({
+        id: cardState.id,
```
CardState is for the react credit cards library right? so `id` is not necessary?

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CardModalInner.js:
```diff
@@ -0,0 +1,179 @@
+            expiration_date: {
+                month: 1,
+                year: 2020
+            },
```
even if you will use this initialValues keep month and year empty

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CardModalInner.js:
```diff
@@ -0,0 +1,179 @@
+                <Input
+                    onFocus={handleInputFocus}
+                    formatter={value => console.log(value)}
+                    name="number"
```
what is this formatter thing? I believe you were just debugging smthg here right?

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CardModalInner.js:
```diff
@@ -0,0 +1,179 @@
+                    <Form.Item
+                        name={['expiration_date', 'month']}
+                        noStyle
+                        rules={[
+                            
+                        ]}>
```
why is the rules array empty?

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCard.js:
```diff
@@ -5,57 +5,75 @@ import cls from 'classnames'
+        try {
+            setIsLoading(true)
+            await sleep(2000)
+            setIsLoading(false)
```
same comment with the other file

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -39,18 +61,21 @@ export const CreditCardList = () => {
+                        grid={{ gutter: 0 }}
+                        dataSource={cardList}
+                        renderItem={card => (
+                            <List.Item>
+                                <CreditCard card={card} onCardInfoChange={onCardInfoChange} selected={selectedCard === card.id} onSelect={onCardSelect} />
+                            </List.Item>
```
remove onSelect from CreditCard since it is not its responsiblity, move the selection logic to List.Item instead.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCard.js:
```diff
@@ -5,57 +5,75 @@ import cls from 'classnames'
     return (
         <div
-            onClick={() => onSelect(cardId)}
+            onClick={() => onSelect(card?.id)}
```
remove this line 36

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCard.js:
```diff
@@ -5,57 +5,75 @@ import cls from 'classnames'
+                        type="radio"
+                        name="selectedCard"
+                        checked={selected}
+                        value={card?.id}
```
remove all the unnecessary null coalescing operators.

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCard.js:
```diff
@@ -5,57 +5,75 @@ import cls from 'classnames'
-                        cancelText="No">
-                        <DeleteOutlined />
-                    </Popconfirm>
+                    &nbsp;{card?.name}
```
I imagined the radio button to be in the list instead. I think thats where it should be. could you move this logic to the upper list component?

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -1,36 +1,58 @@
+        try {
+            setIsLoading(true)
+            await sleep(2000)
+            setIsLoading(false)
```
same comment

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -1,36 +1,58 @@
+        }
+    }
+
+    useEffect(() => getCardList(''), [])
```
do not use this form, use with () => {} because it returns a promise to the useEffect and it will have unintended effects

MehdiSaffar replied:
also use the userid from appcontext. when using userId, also add userId to dependency list of useEffect

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/CreditCardList.js:
```diff
@@ -1,36 +1,58 @@
+
+    const onCardSelect = cardId => setSelectedCard(cardId)
+
+    const onCardInfoChange = async () => getCardList('')
```
also here use the userid

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -1,157 +1,48 @@
-    const handleInputFocus = val => setFocused(val.target.name)
+    const onEdit = async () => {
+        try {
+            setIsLoading(true)
+            await sleep(2000)
+            setIsLoading(false)
+            const fields = await form.validateFields()
```
same comment

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -1,157 +1,48 @@
+            setIsLoading(false)
+            const fields = await form.validateFields()
+            console.log(fields)
+            onSuccessfulEdit();
+            onCancel();
+            notification.success({ message: 'You have successfully changed your credit card information!' })
```
same comment, use only the first callback

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/EditCardModal.js:
```diff
@@ -1,157 +1,48 @@
+            onOk={onEdit}
+            onCancel={onCancel}
             cancelText="Go Back"
             okText="Edit">
```
i think you might need to use forceRender here because of the warning from Antd in Form or Modal section I dont remember.

MehdiSaffar commented on ecommerce/frontend/src/utils.js:
```diff
@@ -70,6 +70,29 @@ export function sleep(ms) {
+    return {
+        name: owner_name,
+        number: serial_number,
+        expiry: (month <= 9 ? `0${month}` : month) + '/' + (year % 1000),
```
for month two comments:
 - sometimes january is counted as 0 sometimes 1, make sure of that
 - use `month.toString().padLeft(2, '0')` instead 

shouldn't year be year % 100 instead?

MehdiSaffar reviewed (changes requested):
Really like the work, just a few nitpicks.

Suggestions:
 - if possible make CVV field small (very optional)
 - when presented with multiple cards at the beginning, select the first one (or maybe default card id from backend?)
 - add more draft code with app context for user information
 - the radio button needs id as a number (just like in the card object), right now it doesnt work.
 - add focus to card name (using owner_name for example) so that the card goes back to front

Commit:
Completed the final changes
- AddCardModel and EditCardModal has been merged into single component CreditCardModal
- Minor changes here and there
- Now the first card in the list is selected by default
